design_docs/oscode/oscode_sundials.md

```diff
@@ -0,0 +1,375 @@
+
+At each step, either the ARDC or RK solver is chosen for that step based on the relative difference in error between the two methods. For RK, the error is generated by the difference in the $N$th and $N-1$th order RK steps. The ARDC steps error for the function evaluation and the first derivative of $x$ with respect to time is defined as
+
+$$
+\delta x_{ARDC} = A_+ \delta f_+ + A_- \delta f_-
+$$
+$$
+\delta f_\pm = f_\pm \sum_{i=0}^n \delta[S_i]_t^{t + h}
+$$
+
+$$
+\delta \dot{x}\_{ARDC} = B\_+ \delta \dot{f}\_+ + B\_- \delta \dot{f}\_-
+$$
+
+$$
+\delta \dot{f}\_\pm = \delta f\_\pm \frac{\dot{f}\_\pm}{f\_\pm}
+$$
+
+OSCODE's time stepping modules may be run in a variety of "modes":
+
+- NORMAL: The solver will take internal steps until it has just overtaken a user-specified output time, $t_{out}$, in the direction of integration, i.e. $t_{n-1} < t_{out} \le t_n$ for forward integration, or $t_{n} < t_{out} \le t_{n-1}$ for backward integration. It will then compute an approximation to the solution $y(t_{out})$ by interpolation (using one of the dense output routines described in the section TBD).
+
+- ONE-STEP: The solver will only take a single internal step $y_{n-1} \rightarrow y_n$ and then return control back to the calling program. If this step will overtake $t_{out}$ then the solver will again return an interpolated result; otherwise it will return a copy of the internal solution $y_n$
+
+- NORMAL-TSTOP: The solver will take internal steps until the next step will overtake $t_{out}$. It will then limit this next step so that $t_n = t_{n-1} + h_n = t_{out}$, and once the step completes it will return a copy of the internal solution
+
+- ONE-STEP-TSTOP: The solver will check whether the next step will overtake $t_{out}$ if not then this mode is identical to "one-step" above; otherwise it will limit this next step so that $t_n = t_{n-1} + h_n = t_{out}$. In either case, once the step completes it will return a copy of the internal solution $y_n$.
+.
+
+We note that interpolated solutions may be slightly less accurate than the internal solutions produced by the solver. Hence, to ensure that the returned value has full method accuracy one of the "tstp" modes may be used.
+
+### Interpolation
+
+As mentioned above, the time-stepping modules in OSCODE support interpolation of solution $y(t_{out})$ and derivatives $y^{(d)}(t_{out})$, where $t_{out}$ occurs within a completed time step from $t_{n-1} \rightarrow t_n$. Additionally, this module supports extrapolation of solutions and derivatives for outside this interval (e.g. to construct predictors for iterative nonlinear and linear solvers). To this end, OSCODE currently supports construction of polynomial interpolants $p_q(t)$ of polynomial degree up to $q=5$, although users may select interpolants of lower degree.
```

Comment:
The order of interpolation is going to be determined by different factors depending on which method the step we're interpolation within was taken with. E.g. If it's a RK step, it'll almost certainly be the same order (or one lower) than that of the RK solver. If it was taken with ARDC, it's set freely by the user, but the same order will be used to perform derivatives, integration within ARDC steps. Maybe this is just a bit too complicated for this white paper.

```suggestion
As mentioned above, the time-stepping modules in OSCODE support interpolation of solution $\mathbf{y}(t_{\mathrm{out}})$ and derivatives $\frac{\mathrm{d}\mathbf{y}}{\mathrm{d}t}|_{t = t_{\mathrm{out}}}$, where $t_{\mathrm{out}}$ occurs within a completed time step from $[t_{n-1},  t_n]$. To this end, OSCODE currently supports construction of polynomial interpolants $p_q(t)$ where the degree $q$ is determined by a parameter passed to the solver by the user.
```